### ISSUE TITLE
Apply Post Content Restriction to get_adjacent_post previous/next links

### ISSUE THREAD

**Issue by TonyGravagno** [ultimatemember/ultimatemember#1365]
Add a content restriction to a Post. For example, allow Contributors to see it, but not Subscribers. In that post's template, add Next and Previous Post blocks. Now with a user with a Subscriber role open the previously created post.   
The Next link has the restricted post - I prefer that it would not.

The code to manage this is almost in place:

```php
add_filter( 'get_next_post_where', array( &$this, 'exclude_navigation_posts' ), 99, 5 );
add_filter( 'get_previous_post_where', array( &$this, 'exclude_navigation_posts' ), 99, 5 );
```

[exclude_navigation_posts](https://github.com/ultimatemember/ultimatemember/blob/19c8f62c5857095ae374980e044c7ab482443557/includes/core/class-access.php#L647
)

That code correctly hooks the [$where filter](https://github.com/WordPress/wordpress-develop/blob/1029dde1bc8b8c94dd502c294de51c41c79f741a/src/wp-includes/link-template.php#L1967) from [get_adjacent_post](https://github.com/WordPress/wordpress-develop/blob/1029dde1bc8b8c94dd502c294de51c41c79f741a/src/wp-includes/link-template.php#L1814-L2018). 

However, exclude_navigation_posts is only processing terms. It's not using the post-level content restriction.
Maybe I'm missing something?
Thanks!

WordPress 6.4.1 with UM 2.3.4.

![image](https://github.com/ultimatemember/ultimatemember/assets/499473/6495e149-99dd-436f-864c-673ef4240f4f)
![image](https://github.com/ultimatemember/ultimatemember/assets/499473/538076d2-bd34-4607-8b81-5416383307db)

**Comment by TonyGravagno** [comment]
I see what the issue is here...  

When "Would you like to display 404 page when users haven't access?" is UNchecked:

- A post shows in the adjacent post links.
- The access restricted message shows when the post is loaded.

When Checked: (This is the desired behavior!)

- A post does NOT show in the adjacent post links.
- Access to the page results in a 404.

The connection is not apparent between showing a 404 and showing the post in the adjacent links.  
I hope this can be made more clear in the docs and FAQ. And that we can see hook code in the doc for refining how this behaves.

![image](https://github.com/ultimatemember/ultimatemember/assets/499473/f73d154f-7b29-4ddb-a425-1604c86f8c59)

**Comment by TonyGravagno** [comment]
Final note (?) : OK, I see that the same rules apply to taxonomy, where the 404 toggle is in the taxonomy configuration. The documentation is very clear about this, I just didn't "get it" initially. This is not about "displaying a 404" page, it's about excluding terms and restricted posts from WP_Query, which is clarified in the field description rather than the label.

![image](https://github.com/ultimatemember/ultimatemember/assets/499473/0f8b1c8d-464d-4862-910f-374df1c8fcb0)

In this experience I've learned something that is very important but I didn't "get it" on this initially either. Post-level content restriction is processed first, and if it is present, taxonomy/terms restrictions are not processed. That is, if you restrict access to a post to a specific role, there is not another check to see if the taxonomy. At first I thought this was a bug and then it made perfect sense:

- If you restrict a post to Contributors, then all contributors get access and you don't need to check the taxonomy. With the 404 box unchecked, everyone sees links and the post template, but not content. With the 404 box checked, the post is completely hidden from everyone except the checked user roles.
- If you use taxonomy, that controls the access - you don't need to use the post-level access restriction too. In fact, if you want to control access with taxonomy, you must disable "Restrict access to post?". This results in the "404" behavior, where WP_QUERY excludes the taxonomy terms from all queries, making the post disappear from all views for unauthorized eyes.

![image](https://github.com/ultimatemember/ultimatemember/assets/499473/e54fabbb-7cd0-418a-be8a-cf8bf18bdfd1)

Understanding these rules took a couple days, during which I wrote a lot of code to accommodate what I thought were bugs or missing functionality. Now I can archive all of that code, because this awesome software already does everything that is desired in this area ... even if it's not completely clear exactly how this feature works. ;)
